Proofs/Experiments/Max/MaxTandem.lean

```diff
@@ -0,0 +1,899 @@
+
+/--
+info: 'MaxTandem.program.stepi_0x8bc_cut' depends on axioms: [propext, Classical.choice, Quot.sound]
+-/
+#guard_msgs in #print axioms program.stepi_0x8bc_cut
+
+--- 12/15: str  w0, [sp, #28]
+theorem program.stepi_0x8c0_cut (s sn : ArmState)
+  (h_program : s.program = program)
+  (h_pc : r StateField.PC s = 0x8c0#64)
+  (h_err : r StateField.ERR s = StateError.None)
+  (h_sp_aligned : CheckSPAlignment s)
+  (h_step : sn = run 1 s) :
+  cut sn = false ∧
+  r StateField.PC sn = 0x8c4#64 ∧
+  r StateField.ERR sn = .None ∧
+  sn.program = program ∧
+  sn[sn.sp + 28, 4] = s.x0.zeroExtend 32  ∧
+  sn.sp = s.sp ∧
+  CheckSPAlignment sn :=  by
+  have := program.stepi_eq_0x8c0 h_program h_pc h_err
+  simp only [minimal_theory] at this
+  simp_all only [run, cut, this, state_simp_rules, bitvec_rules, minimal_theory]
+  simp [pcs]
+
+  /--
+info: 'MaxTandem.program.stepi_0x8c0_cut' depends on axioms: [propext, Classical.choice, Lean.ofReduceBool, Quot.sound]
+-/
+#guard_msgs in #print axioms program.stepi_0x8c0_cut
+
+-- 13/15: ldr  w0, [sp, #28]
+theorem program.stepi_0x8c4_cut (s sn : ArmState)
+  (h_program : s.program = program)
+  (h_pc : r StateField.PC s = 0x8c4#64)
+  (h_err : r StateField.ERR s = StateError.None)
+  (h_sp_aligned : CheckSPAlignment s)
+  (h_step : sn = run 1 s) :
+  r StateField.PC sn = 0x8c8#64 ∧
+  r StateField.ERR sn = .None ∧
+  sn.program = program ∧
+  sn.mem = s.mem ∧
+  cut sn = false ∧
+  sn.sp = s.sp ∧
+  sn[sn.sp + 28#64, 4] = s[s.sp + 28#64, 4]  ∧
+  sn.x0 = BitVec.zeroExtend 64 (sn[sn.sp + 28#64, 4])  ∧
+  CheckSPAlignment sn :=  by
+  have := program.stepi_eq_0x8c4 h_program h_pc h_err
+  simp only [minimal_theory] at this
+  simp_all only [run, cut, this, state_simp_rules, bitvec_rules, minimal_theory]
+  simp [pcs]
+
+/--
+info: 'MaxTandem.program.stepi_0x8c4_cut' depends on axioms: [propext, Classical.choice, Quot.sound]
+-/
+#guard_msgs in #print axioms program.stepi_0x8c4_cut
+
+-- 14/15: add  sp, sp, #0x20
+theorem program.stepi_0x8c8_cut (s sn : ArmState)
+  (h_program : s.program = program)
+  (h_pc : r StateField.PC s = 0x8c8#64)
+  (h_err : r StateField.ERR s = StateError.None)
+  (h_sp_aligned : CheckSPAlignment s)
+  (h_step : sn = run 1 s) :
+  r StateField.PC sn = 0x8cc#64 ∧
+  r StateField.ERR sn = .None ∧
+  cut sn = true ∧
+  sn.program = program ∧
+  sn.x0 = s.x0 ∧
+  sn.sp = s.sp + 0x20#64 ∧
+  CheckSPAlignment sn :=  by
+  have := program.stepi_eq_0x8c8 h_program h_pc h_err
+  simp only [minimal_theory] at this
+  simp_all only [run, cut, this, state_simp_rules, bitvec_rules, minimal_theory]
+  simp (config := {ground := true, decide := true})
+  simp only [fst_AddWithCarry_eq_add, true_and]
+  apply Aligned_BitVecAdd_64_4
+  · assumption
+  · decide
+
+/--
+info: 'MaxTandem.program.stepi_0x8c8_cut' depends on axioms: [propext, Classical.choice, Lean.ofReduceBool, Quot.sound]
+-/
+#guard_msgs in #print axioms program.stepi_0x8c8_cut
+
+end CutTheorems
+
+/--
```

Comment:
I wonder if these tactics should be moved somewhere else, like in Tactics.

Reply:
I wasn't sure, because these felt specialized to this particular proof style, which will change when @alexkeizer merges in the shiny new symbolic simulation :) I'd be happy to move them to a separate file when we use these tactics multiple times. Does that sound fair to you?